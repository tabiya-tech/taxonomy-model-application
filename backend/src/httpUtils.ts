import {APIGatewayProxyResult} from "aws-lambda/trigger/api-gateway-proxy";

export enum HTTP_VERBS {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  PATCH = 'PATCH',
  OPTIONS = 'OPTIONS',
}

export enum ReasonPhrases {
  INTERNAL_SERVER_ERROR = "Internal Server Error",
  METHOD_NOT_ALLOWED = "Method Not Allowed",
  NOT_FOUND = "Not Found"
}

export enum StatusCodes {
  INTERNAL_SERVER_ERROR = 500,
  METHOD_NOT_ALLOWED = 405,
  NOT_FOUND = 404,
  OK = 200,
}

export enum ErrorCodes {
  METHOD_NOT_ALLOWED = "METHOD_NOT_ALLOWED",
  NOT_FOUND = "NOT_FOUND",
  INTERNAL_SERVER_ERROR = "INTERNAL_SERVER_ERROR",
}

// See https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format
// For the format of the return value
export function response(statusCode: StatusCodes, body: object | string): APIGatewayProxyResult {
  return {
    isBase64Encoded: false,
    headers: {
      "Access-Control-Allow-Origin": "*", // Required for CORS support to work
    },
    multiValueHeaders: {},
    statusCode: statusCode,
    body: (typeof body === "object" ? JSON.stringify(body) : body)
  }
}

export type ResponseError = {
  error: string, // This is what the UI should use to display some useful information
  message: string, // the error message offers better developer experience. UI should not display this message.
  details: string, // this may be some cryptic message only useful to a developer
  path: string // this is the path to the resource, if it exists
};

export function errorResponse(statusCode: StatusCodes, error: string, message: string, details: string, path: string): APIGatewayProxyResult {
  return response(statusCode, [{
    error: error,
    message: message,
    details: details,
    path: path
  }]);
}

export function errorsResponse(statusCode: StatusCodes, errors: ResponseError[]): APIGatewayProxyResult {
  return response(statusCode, errors);
}

// Standard error responses used repeatedly
export const STD_ERRORS_RESPONSES = {
  METHOD_NOT_ALLOWED: errorResponse(StatusCodes.METHOD_NOT_ALLOWED, ErrorCodes.METHOD_NOT_ALLOWED, ReasonPhrases.METHOD_NOT_ALLOWED, "", ""),
  NOT_FOUND: errorResponse(StatusCodes.NOT_FOUND, ErrorCodes.NOT_FOUND, ReasonPhrases.NOT_FOUND, "", ""),
  INTERNAL_SERVER_ERROR: errorResponse(StatusCodes.INTERNAL_SERVER_ERROR, ErrorCodes.INTERNAL_SERVER_ERROR, ReasonPhrases.INTERNAL_SERVER_ERROR, "", ""),
}