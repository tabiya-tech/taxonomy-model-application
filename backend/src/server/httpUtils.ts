import { APIGatewayProxyResult } from "aws-lambda/trigger/api-gateway-proxy";
import ErrorAPISpecs from "api-specifications/error";
import ModelInfoAPISpecs from "api-specifications/modelInfo";
import ImportAPISpecs from "api-specifications/import";

export enum HTTP_VERBS {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  DELETE = "DELETE",
  PATCH = "PATCH",
  OPTIONS = "OPTIONS",
}

export enum StatusCodes {
  INTERNAL_SERVER_ERROR = 500,
  BAD_REQUEST = 400,
  METHOD_NOT_ALLOWED = 405,
  NOT_FOUND = 404,
  UNSUPPORTED_MEDIA_TYPE = 415,
  TOO_LARGE_PAYLOAD = 413,
  OK = 200,
  CREATED = 201,
  ACCEPTED = 202,
}

// See https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format
// For the format of the return value
export function response(
  statusCode: StatusCodes,
  body: object | string,
  headers?: {
    [key: string]: string;
  }
): APIGatewayProxyResult {
  return {
    isBase64Encoded: false,
    headers: {
      ...(headers ?? {}),
      "Access-Control-Allow-Origin": "*", // Required for CORS support to work
    },
    multiValueHeaders: {},
    statusCode: statusCode,
    body: typeof body === "object" ? JSON.stringify(body) : body,
  };
}

export function responseJSON(statusCode: StatusCodes, body: object | string): APIGatewayProxyResult {
  return response(statusCode, body, { "Content-Type": "application/json" });
}

function _errorResponse(statusCode: StatusCodes, error: ErrorAPISpecs.Types.Payload): APIGatewayProxyResult {
  return response(statusCode, error);
}

export function errorResponse(
  statusCode: StatusCodes,
  errorCode:
    | ErrorAPISpecs.Constants.ErrorCodes
    | ImportAPISpecs.Enums.POST.Response.ImportResponseErrorCodes
    | ModelInfoAPISpecs.Enums.POST.Response.ErrorCodes
    | ModelInfoAPISpecs.Enums.GET.Response.ErrorCodes,
  message: string,
  details: string
): APIGatewayProxyResult {
  return _errorResponse(statusCode, {
    errorCode: errorCode,
    message: message ?? "",
    details: details ?? "",
  });
}

// Standard error responses used repeatedly
export const STD_ERRORS_RESPONSES = {
  METHOD_NOT_ALLOWED: errorResponse(
    StatusCodes.METHOD_NOT_ALLOWED,
    ErrorAPISpecs.Constants.ErrorCodes.METHOD_NOT_ALLOWED,
    ErrorAPISpecs.Constants.ReasonPhrases.METHOD_NOT_ALLOWED,
    ""
  ),
  NOT_FOUND: errorResponse(
    StatusCodes.NOT_FOUND,
    ErrorAPISpecs.Constants.ErrorCodes.NOT_FOUND,
    ErrorAPISpecs.Constants.ReasonPhrases.NOT_FOUND,
    ""
  ),
  INTERNAL_SERVER_ERROR: errorResponse(
    StatusCodes.INTERNAL_SERVER_ERROR,
    ErrorAPISpecs.Constants.ErrorCodes.INTERNAL_SERVER_ERROR,
    ErrorAPISpecs.Constants.ReasonPhrases.INTERNAL_SERVER_ERROR,
    ""
  ),
  MALFORMED_BODY_ERROR: (errorDetails: string) =>
    errorResponse(
      StatusCodes.BAD_REQUEST,
      ErrorAPISpecs.Constants.ErrorCodes.MALFORMED_BODY,
      ErrorAPISpecs.Constants.ReasonPhrases.MALFORMED_BODY,
      errorDetails
    ),
  INVALID_JSON_SCHEMA_ERROR: (errorDetails: string) =>
    errorResponse(
      StatusCodes.BAD_REQUEST,
      ErrorAPISpecs.Constants.ErrorCodes.INVALID_JSON_SCHEMA,
      ErrorAPISpecs.Constants.ReasonPhrases.INVALID_JSON_SCHEMA,
      errorDetails
    ),
  UNSUPPORTED_MEDIA_TYPE_ERROR: errorResponse(
    StatusCodes.UNSUPPORTED_MEDIA_TYPE,
    ErrorAPISpecs.Constants.ErrorCodes.UNSUPPORTED_MEDIA_TYPE,
    ErrorAPISpecs.Constants.ReasonPhrases.UNSUPPORTED_MEDIA_TYPE,
    "Content-Type should be application/json"
  ),
  TOO_LARGE_PAYLOAD_ERROR: (errorDetails: string) =>
    errorResponse(
      StatusCodes.TOO_LARGE_PAYLOAD,
      ErrorAPISpecs.Constants.ErrorCodes.TOO_LARGE_PAYLOAD,
      ErrorAPISpecs.Constants.ReasonPhrases.TOO_LARGE_PAYLOAD,
      errorDetails
    ),
};

export function redactCredentialsFromURI(uri: string) {
  // Regular expression pattern to match username and password
  const pattern = /\/\/[^@]+@/;

  // Replace the matched username and password with asterisks
  return uri.replace(pattern, "//*:*@");
}
