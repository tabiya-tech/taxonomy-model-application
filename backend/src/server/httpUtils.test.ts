import Ajv from 'ajv';
import addFormats from "ajv-formats";
import 'jest-performance-matchers';
import {errorResponse, redactCredentialsFromURI, response, STD_ERRORS_RESPONSES} from "./httpUtils";

import APIError from "api-specifications/error";

describe("test response function", () => {

  test("should return correct response for an object", () => {
    // GIVEN a status code
    const givenStatusCode = 200;
    // AND a javascript object
    const givenObject = {bar: "baz"};

    //WHEN response is invoked for the given status javascript object
    const result = response(givenStatusCode, givenObject);

    //THEN expect statusCode to be
    expect(result.statusCode).toEqual(givenStatusCode);
    // AND body to be a json string of the given object
    expect(result.body).toEqual(JSON.stringify(givenObject));
    // AND CORS headers to be set
    expect(result.headers).toEqual({"Access-Control-Allow-Origin": "*"});
    // AND isBase64Encoded to be false
    expect(result.isBase64Encoded).toEqual(false);
  });

  test("should return correct response for a string", () => {
    // GIVEN a status code
    const givenStatusCode = 200;
    // AND a string
    const givenString = JSON.stringify({bar: "baz"});

    //WHEN response is invoked for the given status and string
    const result = response(givenStatusCode, givenString);

    //THEN expect statusCode to be
    expect(result.statusCode).toEqual(givenStatusCode);
    // AND body to be the  given string
    expect(result.body).toEqual(givenString);
    // AND CORS headers to be set
    expect(result.headers).toEqual({"Access-Control-Allow-Origin": "*"});
    // AND isBase64Encoded to be false
    expect(result.isBase64Encoded).toEqual(false);
  });
});

describe("test the errorResponse function", () => {
  test("should return correct response for an error", () => {
    // GIVEN a status code
    const givenStatusCode = 500;
    // AND some error details
    const givenError: APIError.POST.Response.Payload = {
      errorCode: APIError.Constants.ErrorCodes.INTERNAL_SERVER_ERROR,
      message: "message",
      details: "details"
    };

    //WHEN errorResponse is invoked for the given status and error
    const result = errorResponse(givenStatusCode, givenError.errorCode, givenError.message, givenError.details);

    //THEN expect statusCode to be
    expect(result.statusCode).toEqual(givenStatusCode);
    // AND body to be a json string of an array than contains given error
    expect(result.body).toEqual(JSON.stringify(givenError));
    // AND a body should validate against the APIError schema
    const ajv = new Ajv({validateSchema: true, strict: true, allErrors: true});
    addFormats(ajv);
    const validateResponse = ajv.compile(APIError.POST.Response.Schema);
    validateResponse(JSON.parse(result.body));
    expect(validateResponse.errors).toBeNull();
  });
  test.each([
    ["undefined", undefined],
    ["null", null],
  ])
  ("should return correct response for an error even is some properties are %s", (description, value) => {
    // GIVEN a status code
    const givenStatusCode = 500;
    // AND some error details
    const givenError: APIError.POST.Response.Payload = {
      errorCode: APIError.Constants.ErrorCodes.INTERNAL_SERVER_ERROR,
      // @ts-ignore
      message: value,
      // @ts-ignore
      details: value
    };

    //WHEN errorResponse is invoked for the given status and error
    const result = errorResponse(givenStatusCode, givenError.errorCode, givenError.message, givenError.details);

    //THEN expect statusCode to be
    expect(result.statusCode).toEqual(givenStatusCode);
    // AND body to be a json string of an array than contains given error
    expect(result.body).toEqual(JSON.stringify({...givenError, message: "", details: ""}));
    // AND a body should validate against the APIError schema
    const ajv = new Ajv({validateSchema: true, strict: true, allErrors: true});
    addFormats(ajv);
    const validateResponse = ajv.compile(APIError.POST.Response.Schema);
    validateResponse(JSON.parse(result.body));
    expect(validateResponse.errors).toBeNull();
  });

});

describe("test the STD_ERRORS_RESPONSES", () => {
  test("STD_ERRORS_RESPONSES.METHOD_NOT_ALLOWED", () => {
    expect(STD_ERRORS_RESPONSES.METHOD_NOT_ALLOWED).toMatchSnapshot();
  });

  test("STD_ERRORS_RESPONSES.NOT_FOUND", () => {
    expect(STD_ERRORS_RESPONSES.NOT_FOUND).toMatchSnapshot();
  });

  test("STD_ERRORS_RESPONSES.INTERNAL_SERVER_ERROR", () => {
    expect(STD_ERRORS_RESPONSES.INTERNAL_SERVER_ERROR).toMatchSnapshot();
  });
});

describe("test the redactCredentialsFromURI function", () => {

  describe("test typical uri", () => {
    test.each(
      [
        ["http://foo.bar"],
        ["http://foo/bar"],
        ["http://foo/bar?baz=qux"],
        ["https://example.com:8080"],
        ["mongodb://example.com:8080/"], // NOSONAR
      ]
    )
    ("should return the same URI %s if no credentials are present", (uri: string) => {
      const result = redactCredentialsFromURI(uri);
      expect(result).toEqual(uri);
    });

    test.each(
      [
        ["http://user:password@foo.bar", "http://*:*@foo.bar"],
        ["http://user:password@foo/bar", "http://*:*@foo/bar"],
        ["http://user:password@foo/bar?baz=qux", "http://*:*@foo/bar?baz=qux"],
        ["https://user:password@example.com:8080", "https://*:*@example.com:8080"],
        ["mongodb://user:password@example.com:8080/", "mongodb://*:*@example.com:8080/"], // NOSONAR
      ]
    )("should return the redacted URI %s if credentials are present", (uriWithCredentials, uriWithoutCredential) => {
      const result = redactCredentialsFromURI(uriWithCredentials);
      expect(result).toEqual(uriWithoutCredential);
    });
  });
  describe("test credentials are malformed", () => {
    test.each(
      [
        ["//user:password@foo.bar", "//*:*@foo.bar"],
        ["://user:password@foo/bar", "://*:*@foo/bar"],
        ["http://user:@foo/bar?baz=qux", "http://*:*@foo/bar?baz=qux"],
        ["https://user@example.com:8080", "https://*:*@example.com:8080"],
        ["mongodb://:password@example.com:8080/", "mongodb://*:*@example.com:8080/"], // NOSONAR
        ["mongodb://user:password@user:password@example.com:8080/", "mongodb://*:*@example.com:8080/"], // NOSONAR
      ]
    )("should return the redacted URI %s if credentials are malformed", (uriWithCredentials, uriWithoutCredential) => {
      const result = redactCredentialsFromURI(uriWithCredentials);
      expect(result).toEqual(uriWithoutCredential);
    });
  });

  describe("test function performance", () => {
    const PERF_DURATION = 10;
    test.each([
      ["plain http", "http://foo/bar?baz=qux"],
      ["http with credentials", "http://username:password@foo/bar?baz=qux"],
      ["only username", "http://username:@foo/bar?baz=qux"],
      ["only password", "http://:@foo/bar?baz=qux"],
      ["(extreme long) plain http", "http://foo/bar?baz=" + "qux".repeat(65535)],
      ["(extreme long) with credentials ", "http://" + "username".repeat(32000) + ":" + "password".repeat(32000) + "@foo/bar?baz=" + "qux".repeat(65535)],
      ["(extreme long) only username ", "http://" + "username".repeat(32000) + "@foo/bar?baz=" + "qux".repeat(65535)],
      ["(extreme long) only password ", "http://:" + "username".repeat(32000) + "@foo/bar?baz=" + "qux".repeat(65535)],
      ["(extreme long) with multiple user name passwords", "http://:" + "username:password".repeat(32000) + "@foo/bar?baz=" + "qux".repeat(65535)],
    ])(`It performs fast (<=${PERF_DURATION}ms) and does not hang/cause catastrophic backtracking for '%s'`, (description, uri) => {
      expect(() => {
        redactCredentialsFromURI(uri);
      }).toCompleteWithinQuantile(PERF_DURATION, {iterations: 10, quantile: 90});
    });
  });
});