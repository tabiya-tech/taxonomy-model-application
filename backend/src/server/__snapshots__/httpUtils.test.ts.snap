// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the STD_ERRORS_RESPONSES STD_ERRORS_RESPONSES.INTERNAL_SERVER_ERROR 1`] = `
{
  "body": "{"errorCode":"INTERNAL_SERVER_ERROR","message":"Internal Server Error","details":""}",
  "headers": {
    "Access-Control-Allow-Origin": "https://test-url",
  },
  "isBase64Encoded": false,
  "multiValueHeaders": {},
  "statusCode": 500,
}
`;

exports[`test the STD_ERRORS_RESPONSES STD_ERRORS_RESPONSES.METHOD_NOT_ALLOWED 1`] = `
{
  "body": "{"errorCode":"METHOD_NOT_ALLOWED","message":"Method Not Allowed","details":""}",
  "headers": {
    "Access-Control-Allow-Origin": "https://test-url",
  },
  "isBase64Encoded": false,
  "multiValueHeaders": {},
  "statusCode": 405,
}
`;

exports[`test the STD_ERRORS_RESPONSES STD_ERRORS_RESPONSES.NOT_FOUND 1`] = `
{
  "body": "{"errorCode":"NOT_FOUND","message":"Not Found","details":""}",
  "headers": {
    "Access-Control-Allow-Origin": "https://test-url",
  },
  "isBase64Encoded": false,
  "multiValueHeaders": {},
  "statusCode": 404,
}
`;
